src/azstoragetorch/io.py

```diff
@@ -0,0 +1,188 @@
+        # performance implications as one read() will result in at least one HTTP GET request for each
+        # read() call.
+        #
+        # Instead of allowing the default implementation and having users impacted by these
+        # implications, we raise a NotImplementedError so that readline() and any other methods
+        # that are automatically implemented from readline() (e.g. readlines() and __next__()) can't
+        # be used until we implement a more appropriate strategy.
+        raise NotImplementedError("readline")
+
+    def seek(self, offset: int, whence: int = os.SEEK_SET, /) -> int:
+        self._validate_is_integer("offset", offset)
+        self._validate_is_integer("whence", whence)
+        self._validate_not_closed()
+        return self._seek(offset, whence)
+
+    def seekable(self) -> bool:
+        return self.readable()
+
+    def tell(self) -> int:
+        self._validate_not_closed()
+        return self._position
+
+    def write(self, b: Union[bytes, bytearray], /) -> int:
+        raise NotImplementedError("write")
+
+    def writable(self) -> bool:
+        return False
+
+    def _validate_mode(self, mode: str) -> None:
+        if mode not in get_args(_SUPPORTED_MODES):
+            raise ValueError(f"Unsupported mode: {mode}")
+
+    def _validate_is_integer(self, param_name: str, value: int) -> None:
+        if not isinstance(value, int):
+            raise TypeError(f"{param_name} must be an integer, not: {type(value)}")
+
+    def _validate_min(self, param_name: str, value: int, min_value: int) -> None:
+        if value < min_value:
+            raise ValueError(
+                f"{param_name} must be greater than or equal to {min_value}"
+            )
+
+    def _validate_not_closed(self) -> None:
+        if self.closed:
+            raise ValueError("I/O operation on closed file")
+
+    def _get_sdk_credential(
+        self, blob_url: str, credential: _AZSTORAGETORCH_CREDENTIAL_TYPE
+    ) -> _SDK_CREDENTIAL_TYPE:
+        if credential is False or self._blob_url_has_sas_token(blob_url):
+            return None
+        if credential is None:
+            return DefaultAzureCredential()
+        if isinstance(credential, (AzureSasCredential, TokenCredential)):
+            return credential
+        raise TypeError(f"Unsupported credential: {type(credential)}")
+
+    def _blob_url_has_sas_token(self, blob_url: str) -> bool:
+        parsed_url = urllib.parse.urlparse(blob_url)
+        if parsed_url.query is None:
+            return False
+        parsed_qs = urllib.parse.parse_qs(parsed_url.query)
+        # The signature is always required in a valid SAS token. So look for the "sig"
+        # key to determine if the URL has a SAS token.
+        return "sig" in parsed_qs
+
+    def _get_blob_downloader(self) -> "_BlobDownloader":
+        return _BlobDownloader(self._blob_url, self._sdk_credential)
+
+    def _read(self, size: Optional[int]) -> bytes:
```

Comment:
If we take away the `Optional` in the public `read` we may not need it here either?